aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp

```diff
@@ -276,85 +276,64 @@ std::tuple<Tensor, int64_t> cumulative_and_max_seq_len(Tensor qkv) {
-                          query_offset_differences.cbegin() + 1,
-                          query_offset_differences.cend(),
-                          key_offset_differences.cbegin() + 1) &&
-      std::equal(key_offset_differences.cbegin() + 1,
-                 key_offset_differences.cend(),
-                 value_offset_differences.cbegin() + 1);
-  if (!same_offsets) {
-    return false;
+bool is_safe_to_get_storage_as_tensor(const NestedTensorImpl* tensor) {
+  const auto& tensor_offsets = tensor->get_storage_offsets();
+  const Tensor& tensor_sizes = tensor->get_nested_size_tensor();
+  const Tensor& tensor_strides = tensor->get_nested_stride_tensor();
+
+  const int64_t n_tensors = tensor_strides.size(0);
+  const int64_t n_dims = tensor_strides.size(1);
+
+  if (n_tensors <= 1) {
+    return true;
   }
-  const Tensor& query_strides = query->get_nested_stride_tensor();
-  const Tensor& key_strides = key->get_nested_stride_tensor();
-  const Tensor& value_strides = value->get_nested_stride_tensor();
-
-  const int64_t n_tensors = query_strides.size(0);
-  const int64_t n_dims = query_strides.size(1);
-
-  std::vector<int64_t> previous_q_stride(
-      query_strides.data_ptr<int64_t>(),
-      query_strides.data_ptr<int64_t>() + n_dims);
-  std::vector<int64_t> previous_k_stride(
-      key_strides.data_ptr<int64_t>(),
-      key_strides.data_ptr<int64_t>() + n_dims);
-  std::vector<int64_t> previous_v_stride(
-      value_strides.data_ptr<int64_t>(),
-      value_strides.data_ptr<int64_t>() + n_dims);
 
+  int64_t* previous_tensor_stride = tensor_strides.data_ptr<int64_t>();
   // Check initially that they are in strictly descending order
   for (int i{1}; i < n_dims; i++) {
-    if (previous_q_stride[i - 1] < previous_q_stride[i] ||
-        previous_k_stride[i - 1] < previous_k_stride[i] ||
-        previous_v_stride[i - 1] < previous_v_stride[i]) {
+    if (previous_tensor_stride[i - 1] <= previous_tensor_stride[i]) {
       return false;
     }
   }
+  // Check that each tensor i in the nested tensor has the same strides
+  auto tensor_stride_0 = tensor_strides.stride(0);
 
   for (int i{1}; i < n_tensors; i++) {
     for (const int64_t j : c10::irange(n_dims)) {
-      if (previous_q_stride[j] != query_strides[i][j].item<int64_t>() ||
-          previous_k_stride[j] != key_strides[i][j].item<int64_t>() ||
-          previous_v_stride[j] != value_strides[i][j].item<int64_t>()) {
+      if (previous_tensor_stride[j] !=
+          previous_tensor_stride[i * tensor_stride_0 + j]) {
         return false;
       }
     }
   }
+  // Check the offsets are a constant multiple from the previous numels
+  const int64_t* tensor_size_ptr = tensor_sizes.data_ptr<int64_t>();
+  const int64_t* tensor_stride_ptr = tensor_strides.data_ptr<int64_t>();
+
+  int64_t offset_constant = (tensor_offsets[1] - tensor_offsets[0]) /
+      tensor_size_ptr[0] * tensor_stride_ptr[0];
+
+  for (int64_t i = 2; i < n_tensors; i++) {
```

Comment:
Could probably roll this into the previous loop but this uses the stride matrix and size matrix to compute numel so is based off the first check for descending ordering